lib/ical/recur_iterator.js

```diff
@@ -262,77 +262,79 @@ class RecurIterator {
+        //     would be missed.
+        if (this.has_by_data('BYMONTHDAY')) {
+          this._byDayAndMonthDay(true);
         }
-      }
-      this.last = tempLast.clone();
-
-      //XXX: This feels like a hack, but we need to initialize
-      //     the BYMONTHDAY case correctly and byDayAndMonthDay handles
-      //     this case. It accepts a special flag which will avoid incrementing
-      //     the initial value without the flag days that match the start time
-      //     would be missed.
-      if (this.has_by_data('BYMONTHDAY')) {
-        this._byDayAndMonthDay(true);
-      }
-
-      if (this.last.day > daysInMonth || this.last.day == 0) {
-        throw new Error("Malformed values in BYDAY part");
-      }
-    } else if (this.has_by_data("BYMONTHDAY")) {
-      // Attempting to access `this.last.day` will cause the date to be normalised.
-      // So it will never be a negative value or more than the number of days in the month.
-      // We keep the value in a separate variable instead.
 
-      // Change the day value so that normalisation won't change the month.
-      this.last.day = 1;
-      let daysInMonth = Time.daysInMonth(this.last.month, this.last.year);
+        if (this.last.day > daysInMonth || this.last.day == 0) {
+          throw new Error("Malformed values in BYDAY part");
+        }
+      } else if (this.has_by_data("BYMONTHDAY")) {
+        // Change the day value so that normalisation won't change the month.
+        this.last.day = 1;
 
-      if (dayOffset < 0) {
-        // A negative value represents days before the end of the month.
-        this.last.day = daysInMonth + dayOffset + 1;
-      } else if (this.by_data.BYMONTHDAY[0] > daysInMonth) {
-        // There's no occurrence in this month, find the next valid month.
-        // The longest possible sequence of skipped months is February-April-June,
-        // so we might need to call next_month up to three times.
-        if (!this.next_month() && !this.next_month() && !this.next_month()) {
-          throw new Error("No possible occurrences");
+        // Get a sorted list of days in the starting month that match the rule.
+        let normalized = this.normalizeByMonthDayRules(
+          this.last.year,
+          this.last.month,
+          this.rule.parts.BYMONTHDAY
+        ).filter(d => d >= this.last.day);
+
+        if (normalized.length) {
+          // There's at least one valid day, use it.
+          this.last.day = normalized[0];
+          this.by_data.BYMONTHDAY = normalized;
+        } else {
+          // There's no occurrence in this month, find the next valid month.
+          // The longest possible sequence of skipped months is February-April-June,
+          // so we might need to call next_month up to three times.
+          if (!this.next_month() && !this.next_month() && !this.next_month()) {
+            throw new Error("No possible occurrences");
+          }
         }
```

Comment:
I've thrown out an earlier hack and just used the valid days in the starting month if there is at least one of them.